### PR TITLE
Allow gatekeeper to run in any namespace

### LINKED ISSUE #51: Support installing gatekeeper to any namespace
Currently Gatekeeper only supports installing to the `gatekeeper-system` namespace. That's the default installation namespace that is specified in the Gatekeeper manifests that are imported from the upstream Gatekeeper repo. This issue is to try overriding the namespace in each of the Gatekeeper resources with the namespace being watched by the operator e.g. configured using the `WATCH_NAMESPACE` passed via the downward API. This is because the operator supports running in any namespace and will watch for gatekeeper CRs in that same namespace. Ideally it would be as easy as this to support installing Gatekeeper to any namespace. Otherwise, we may need to make changes to Gatekeeper itself to successfully support this.

### PR DESCRIPTION
**Pull request by ruromero**
Fix #51 

Sets the current namespace to any namespaced object and any object that has a reference to the namespace will be overridden during the reconciliation.

The objects having references to the namespace are the following:
* clusterrolebinding:subject
* rolebinding:subject
* controller-manager:exept-namespace container argument
* validatingWebhook:clientConfig

In case there were multiple namespaces running the gatekeeper operator, the cluster-scoped objects might need to be updated in order to add a new clientConfig to the validatingWebhookConfiguration and more subjects to the clusterRoleBinding.

Signed-off-by: ruromero <rromerom@redhat.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -90,7 +92,8 @@ uninstall: manifests kustomize
 # Deploy controller in the configured Kubernetes cluster in ~/.kube/config
 .PHONY: deploy
 deploy: manifests kustomize
-	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
+	cd config/default && $(KUSTOMIZE) edit set namespace $(NAMESPACE) && cd -;\
```
Comment (font) [inline Makefile:95]: Is `;\` needed here and the next line?

File: controllers/gatekeeper_controller.go [hunk 2]
```diff
@@ -624,3 +635,86 @@ func setContainerArg(obj *unstructured.Unstructured, containerName, argName stri
...
+		}
+	}
+	if err := unstructured.SetNestedSlice(obj.Object, subjects, "subjects"); err != nil {
+		return errors.Wrapf(err, "Failed to set updated subjects to rolebinding")
+	}
+	return setContainerArg(obj, managerContainer, ExemptNamespaceArg, namespace)
+}
+
+// toMap Convenience method to convert any struct into a map
+func toMap(obj interface{}) map[string]interface{} {
```
Comment (font) [inline controllers/gatekeeper_controller.go:703]: These functions below look to be left over from a previous PR.

File: controllers/gatekeeper_controller.go [hunk 3]
```diff
@@ -624,3 +635,86 @@ func setContainerArg(obj *unstructured.Unstructured, containerName, argName stri
...
+		return nil
+	}
+	subjects, found, err := unstructured.NestedSlice(obj.Object, "subjects")
+	if !found || err != nil {
+		return errors.Wrapf(err, "Failed to retrieve subjects from roleBinding")
+	}
+	for _, s := range subjects {
+		subject := s.(map[string]interface{})
+		if err := unstructured.SetNestedField(subject, namespace, "namespace"); err != nil {
+			return errors.Wrapf(err, "Failed to set namespace to rolebinding subject")
```
Comment (font) [inline controllers/gatekeeper_controller.go:693]: s/to/for/

File: controllers/gatekeeper_controller.go [hunk 4]
```diff
@@ -624,3 +635,86 @@ func setContainerArg(obj *unstructured.Unstructured, containerName, argName stri
...
+		}
+	}
+	if err := unstructured.SetNestedSlice(obj.Object, subjects, "subjects"); err != nil {
+		return errors.Wrapf(err, "Failed to set updated subjects to rolebinding")
```
Comment (font) [inline controllers/gatekeeper_controller.go:697]: s/to/in/

File: controllers/gatekeeper_controller.go [hunk 5]
```diff
@@ -624,3 +635,86 @@ func setContainerArg(obj *unstructured.Unstructured, containerName, argName stri
...
+	}
+	return setContainerArg(obj, managerContainer, ExemptNamespaceArg, namespace)
```
Comment (font) [inline controllers/gatekeeper_controller.go:699]: This is left over possibly from a copy/paste typo.

File: controllers/gatekeeper_controller_test.go [hunk 6]
```diff
@@ -48,6 +48,53 @@ func TestDeployValidatingWebhookConfig(t *testing.T) {
...
+	g := NewWithT(t)
+	namespace := "testns"
+	gatekeeper := &operatorv1alpha1.Gatekeeper{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: namespace,
+		},
+	}
+	// Rolebinding namespace overrides
+	rolebindingManifest, err := util.GetManifest(RoleBindingFile)
```
Comment (font) [inline controllers/gatekeeper_controller_test.go:61]: We should check the manifest is not nil.

File: controllers/gatekeeper_controller_test.go [hunk 7]
```diff
@@ -48,6 +48,53 @@ func TestDeployValidatingWebhookConfig(t *testing.T) {
...
+		subject := s.(map[string]interface{})
+		ns, found, err := unstructured.NestedString(subject, "namespace")
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(found).To(BeTrue())
+		g.Expect(ns).To(Equal(namespace))
+	}
+
+	// WebhookFile namespace overrides
+	webhookManifest, err := util.GetManifest(WebhookFile)
+	g.Expect(err).ToNot(HaveOccurred())
```
Comment (font) [inline controllers/gatekeeper_controller_test.go:78]: We should check the manifest is not nil.

File: controllers/gatekeeper_controller_test.go [hunk 8]
```diff
@@ -48,6 +48,53 @@ func TestDeployValidatingWebhookConfig(t *testing.T) {
...
+	err = crOverrides(gatekeeper, WebhookFile, webhookManifest)
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(getContainerArguments(g, managerContainer, webhookManifest)).To(HaveKeyWithValue(ExemptNamespaceArg, namespace))
+
+	// ValidatingWebhookConfiguration namespace overrides
+	webhookConfig, err := util.GetManifest(ValidatingWebhookConfiguration)
```
Comment (font) [inline controllers/gatekeeper_controller_test.go:84]: We should check `err` and that the manifest is not nil.

File: controllers/gatekeeper_controller_test.go [hunk 9]
```diff
@@ -48,6 +48,53 @@ func TestDeployValidatingWebhookConfig(t *testing.T) {
...
+	err = crOverrides(gatekeeper, ValidatingWebhookConfiguration, webhookConfig)
+	g.Expect(err).ToNot(HaveOccurred())
+	webhooks, found, err := unstructured.NestedSlice(webhookConfig.Obj.Object, "webhooks")
```
Comment (font) [inline controllers/gatekeeper_controller_test.go:87]: Check `err` did not occur and possibly that `webhooks` is not nil.

File: controllers/gatekeeper_controller_test.go [hunk 10]
```diff
@@ -48,6 +48,53 @@ func TestDeployValidatingWebhookConfig(t *testing.T) {
 	g.Expect(getStaticAssets(gatekeeper)).NotTo(ContainElement(ValidatingWebhookConfiguration))
 }
 
+func TestCustomNamespace(t *testing.T) {
```
Comment (font) [inline controllers/gatekeeper_controller_test.go:51]: This checks the corner cases and I think that's probably good enough as the other manifests are verified in the e2e tests.

**Review by font — CHANGES_REQUESTED** [review]
@ruromero awesome to see that this works without requiring any modifications to the Gatekeeper code base. :+1:

File: Makefile (hunk 1 above)
Comment (ruromero) [reply Makefile:95]: no, it's not

File: controllers/gatekeeper_controller.go (hunk 2 above)
Comment (ruromero) [reply controllers/gatekeeper_controller.go:703]: Mistake while rebasing, thanks

File: controllers/gatekeeper_controller_test.go [hunk 11]
```diff
@@ -48,6 +48,59 @@ func TestDeployValidatingWebhookConfig(t *testing.T) {
...
+			Namespace: namespace,
+		},
+	}
+	// Rolebinding namespace overrides
+	rolebindingManifest, err := util.GetManifest(RoleBindingFile)
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(rolebindingManifest).ToNot(BeNil())
+	err = crOverrides(gatekeeper, RoleBindingFile, rolebindingManifest)
+	g.Expect(err).ToNot(HaveOccurred())
+	subjects, found, err := unstructured.NestedSlice(rolebindingManifest.Obj.Object, "subjects")
```
Comment (font) [inline controllers/gatekeeper_controller_test.go:66]: Do you think we should check subjects is not `nil`?

File: controllers/gatekeeper_controller.go (hunk 5 above)
Comment (font) [reply controllers/gatekeeper_controller.go:697]: Is this left over?

File: controllers/gatekeeper_controller_test.go (hunk 11 above)
Comment (ruromero) [reply controllers/gatekeeper_controller_test.go:66]: I don't think it can be the case. Subjects is an array of type Subject but we can just add it.

File: controllers/gatekeeper_controller.go (hunk 5 above)
Comment (ruromero) [reply controllers/gatekeeper_controller.go:697]: Yes, sorry

**Comment by ruromero** [comment]
Rebased and applied suggested changes

**Review by font — APPROVED** [review]
